{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ORF finding\n",
    "\n",
    "## Deliverables:\n",
    " - Lab5 notebook with complete inspection cells ( Introduction and Results )\n",
    " - findORFs.py - 50 total points\n",
    " - imports\n",
    "     - sequenceAnalysis (for FastaReader at least )\n",
    "     - your OrfFinder class ( placed in sequenceAnalysis )\n",
    "     \n",
    "## required output files:\n",
    " - tass2ORFdata-ATG-100.txt \n",
    " \n",
    "## classes required (minimally):\n",
    " - OrfFinder\n",
    " - FastAreader\n",
    " \n",
    "## possible extra credit options - 10 additional points\n",
    " - minOrf= [integer]  This allows exclusion of genes with lengths < minOrf  (Note this length includes both start and stop codons)\n",
    " - biggestOrfOnly True | False  This allows reporting of all genes or only the largest within an ORF\n",
    " - startCodons= ['ATG' | 'TTG' | 'GTG']  This parameter produces a list of start codons used in determining genes ( you might convert this to a more useful container like a set.)\n",
    " - stopCodons = ['TAA' | 'TGA' | 'TAG']  This parameter specifies which codons are to be treated as valid stop codons\n",
    " \n",
    "## Due: Monday May 22, 2023 11:59pm\n",
    "\n",
    "# Overview\n",
    "This lab will give you practice with a real algorithm and with formatted output. We will be finding genes in a genome.\n",
    "\n",
    "An open reading frame (ORF) is a region of DNA that is free from stop codons (TAG, TAA, TGA). We can then define a putative gene in that ORF by the presence of a start codon at the 5' end of the ORF and a stop codon at the 3' end. Start codons in some bacteria, archaea and viruses can be ATG, GTG or CTG, and in some cases TTG.  \n",
    "\n",
    "Since genes are made up of codons of three symbols, any particular sequence of DNA can be interpreted in each of 3 overlapping coding frames, and since cellular genomic DNA is double stranded, we can have genes located on both strands, yielding 6 possible coding frames for the many genes in a genome.  Genes can also overlap each other.\n",
    "\n",
    "To limit the number of possible genes that we might find, two basic strategies are sometimes used:\n",
    "\n",
    " - look for genes that are longer than some predefined number (100 bases for example) and\n",
    " - only consider the largest coding region within an ORF, rather than considering all substrings in the ORF that begin with intermediate start codons.\n",
    " \n",
    "A well known ORF finder is available at NCBI - ORF Finder https://www.ncbi.nlm.nih.gov/orffinder/\n",
    "\n",
    "In this assignment, you will write an equivalent tool that provides the same function, though without the graphical display.\n",
    "\n",
    "Use your sequenceAnalysis module in designing a solution. \n",
    "\n",
    "Your design is essential for success.  Spend time considering how best to solve this problem.  Write out pseudocode that describes the core algorithm.  This pseudocode is required to be delivered in comments at the top of your code submission.  Please start early on this assignment, and consider the algorithm carefully.  \n",
    " - Start and stop codons must be in the same frame, \n",
    " - many possible Met codons (or other possible starts) may exist in an ORF, and \n",
    " - genes can have significant overlap.\n",
    " - In some cases, 2 or more ORFS can exist on opposite strands in the same coordinate space.\n",
    "\n",
    "A partial solution exists using regular expressions.  This solution does not handle any of the extra-credit possibilities, and does not handle the required boundary conditions involving gene fragments. Feel free to look, though I did not find this approach helpful. It also uses regular expressions, which are both difficult to read and disallowed for this assignment.\n",
    "\n",
    "Your implementation should be designed to add capability to your sequenceAnalysis module. Please carefully consider where that capability belongs. You are required to make use of your sequenceAnalysis module for most of the code and for FastAreader. You might consider adding  a function to calculate reverse complement. You should also add an OrfFinder class to sequenceAnalysis. For this assignment, you are the designer of your toolbox - choose wisely. You might use this again in your final project.\n",
    "\n",
    "# ORF finder\n",
    "Create a program to analyze a FASTA-formatted file containing a sequence of DNA and find the ORFs (start and stop codons). Your program is called findORFs.py and it:\n",
    "\n",
    " - reads in FASTA-formatted sequence data from a file using STDIN\n",
    " - finds ORFs that are >= 100 nucleotides in length\n",
    " - writes formatted output data to a text file using STDOUT\n",
    " - if there are multiple FASTA-formatted sequences in the file, you need to perform Orf-finding on each of them independently.\n",
    "\n",
    "# Design specification\n",
    "\n",
    "## Input file\n",
    "Input data must make use of STDIN. This will allow commandLine execution using: \n",
    "- findOrfs.py < someInput.\n",
    "I will provide 2 test files for your use.  For debugging purposes, use lab5test.fa, which is provided in Canvas::files.  As your final test, use the tass2.fa genome. I have provided the output generated by my program for your comparison.\n",
    "\n",
    "## Output file\n",
    "\n",
    "Write your tass2 output to a file named: tass2ORFdata-ATG-100.txt using STDOUT.  This file should be generated with the following program features:\n",
    "\n",
    " - The minimum gene size to report is 100 nucleotides (including stop and start)\n",
    " - The only start codon to consider is ATG\n",
    " - The stop codons in DNA form are: TAA TGA TAG\n",
    " - Only the largest putative gene in an ORF is reported\n",
    " - The input file is tass2.fa\n",
    " \n",
    "I have provided the output generated using a minimum gene size of 300, which you can use for testing. \n",
    "\n",
    "## Gene finding\n",
    "\n",
    "Your  program should find genes in each of the six possible coding frames.  The genes that you find will begin with a start codon and end with a stop codon, with the exception of those gene-fragments that might be located on the ends of the sequence (see boundary conditions).\n",
    "\n",
    "## Boundary conditions\n",
    "\n",
    "You may find that an ORF exists at either or both termini of the sequence you are examining.  This may mean that you will only be seeing a gene fragment located at the sequence termini. \n",
    "\n",
    "For example, consider the sequence:<br>\n",
    "AAA AAA AAA TGA CCC CCC ...\n",
    "\n",
    "Here, we see a stop codon with a set of AAA codons located upstream.  In this case, assume that a start codon exists upstream of our sequence, and report this as a gene in Frame 1, starting at position 1, ending at position 12 and having a length of 12.  \n",
    "\n",
    "The same may occur on the 3' end of the sequence, for example: <br>\n",
    "A AAA AAA ATG CCC CCC CCC CC  \n",
    "\n",
    "In this case, report a gene in Frame 2 starting at position 8, ending at position 21 and having a length of 14.  \n",
    "\n",
    "And, of course genes can exist on the opposite strand, so: <br>\n",
    "TTA AAA AAA AAA CAT CC<br>\n",
    "would be reported in Frame -3, starting in position 1, ending at position 15, with length 15.\n",
    "\n",
    "NCBI ORF finder is a good test case to see what your output should look like.  THe NCBI program does not handle this boundary case however.\n",
    "\n",
    "## Sort Order\n",
    "\n",
    "Your output should be sorted by decreasing ORF size, and in cases where multiple ORFs have the same size, sort those by the left position of the gene.  (build a key that makes this easy).\n",
    "\n",
    "## Minimum gene size (option)\n",
    "\n",
    "The minimum gene size to report is 100 nucleotides.  This count of nucleotides includes the start codon and stop codon.\n",
    "\n",
    "## Largest ORF (option)\n",
    "\n",
    "In any ORF, there may be multiple start codons present.  For the default assignment, only consider the largest gene in an ORF. For the extra-credit assignment, if this (largest ORF) option is not set, report all ORFs found (subject to the minimum size option).\n",
    "\n",
    "## Start codon(s) (option)\n",
    "\n",
    "For the default part of the program, only consider ATG as a start codon. For extra credit, use an input parameter that specifies a set of start codons.\n",
    "\n",
    "## Stop codon(s) (option)\n",
    "\n",
    "For the default part of the program, only consider TAG, TAA and TGA as stop codons. For extra credit,  use an input parameter that specifies the set of stop codons.\n",
    "\n",
    "## Report output\n",
    "\n",
    "All positions reported use the coordinate position of the top strand, and the first position in the sequence (on the left) is position 1.  This coordinate system is used for all 6 coding frames.\n",
    "\n",
    "Frames 1 through 3 can be determined using the start position of the gene.  Assuming the position is p, (p mod 3) + 1 will provide the coding frame.  For genes on the bottom strand, the frame is based on the right hand side of the sequence (see below).\n",
    "\n",
    "For each putative gene, your report must provide:\n",
    "\n",
    " - coding frame. for the top strand, a gene starting in the first position in the sequence is in frame 1.  For the bottom strand, frames are numbered based on the right-hand coordinate position of the sequence.  So, if the sequence is 100 bases long, and you have a gene in positions 89..100, that gene is in frame -1. You can still use the coordinate positions, but the calculation is:  - ( (len(sequence) - last position) mod 3 + 1).  Remember that the last position in this case corresponds to the start codon of the gene on the bottom strand.  These conventions make sense if you think about the reverse complement of the original sequence. The coding frame should be reported with a + or - as needed to describe the coding frame\n",
    " - start position. This position should correspond to the first base of the start codon (top strand) or last base of the stop codon (bottom strand), or the terminus of the genome if this is a gene fragment.\n",
    " - end position. This position should correspond to the first base of the start codon (bottom strand) or last base of the stop codon (top strand), or the terminus of the genome if this is a gene fragment.\n",
    " - length. This includes the start and end codon\n",
    " \n",
    "\n",
    "Line 1 - the header line from the original sequence (without >)\n",
    "\n",
    "Lines 2 .... (use '{:+d} {:>5d}..{:>5d} {:>5d}' or equivalent)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "+1 57166..61908  4743 <br>\n",
    "..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Extra Credit - 10 points possible\n",
    " - Allow your program to optionally use a set of start codons - ATG, GTG, TTG for example.\n",
    " - Allow your program to optionally use a set of stop codons - TAG, TGA, TAA for example.\n",
    " - Allow your program to print every putative gene in an ORF instead of only the largest.\n",
    " - Allow your program to have varying ORF size minimums \n",
    "\n",
    "# Submit your code and answers\n",
    "For this lab, you should upload the following three files as attachments:\n",
    "\n",
    " - your sequenceAnalysis.py module\n",
    " - findORFs.py, include design in # comments at the beginning of findORFs.py\n",
    " - tass2ORFdata-ATG-100.txt\n",
    "\n",
    "Important: to get full credit on this lab assignment, each of the code files you submit needs to:\n",
    " - Run properly (execute and produce the correct output)\n",
    " - Include an overview about what your program is designed to do with expected inputs and outputs\n",
    " - Include design (place at the top of your file below your name, group, and description). Include any assumptions or design decisions you made in writing your code\n",
    " - Adhere to the Report format specification\n",
    " - Include docstrings for Program, Module, Class, Methods and Functions\n",
    " - Contain in-line comments\n",
    " \n",
    " Congratulations, you finished your fifth lab assignment!\n",
    "\n",
    "# Hints:\n",
    "## One strand at a time\n",
    "\n",
    "One design solution considers only the top strand to generate gene candidates. We then generate the reverse complement and a second list of gene candidates.  When done in this way, remember to fix the start/end coordinates that you get from the reverse complement solution, since those will be based on the other end of the sequence. \n",
    "\n",
    "An alternative design exists for the reverse strand that does not require generation of the reverse complement. This solution is interesting and may be simpler to consider. This solution scans the bottom strand left to right, finding stop codons first (reverse complement), then looking for start codons (reverse complement).\n",
    "\n",
    "A third solution for the bottom strand would scan right to left, finding starts until a stop is encountered.\n",
    "\n",
    "## Start and Stop handling\n",
    "\n",
    "If you are going to do the extra-credit, then you will need to remember where the starts are until you find a stop.  This is straightforward if you place those start positions in a list, organized by frame. Even if you are not doing the extra-credit, placing starts on a list organized by frame will really simplify your code. This solution would require three lists for each of the three reading frames.\n",
    "\n",
    "As an alternative, you could scan each of the three reading frames seperately, counting by threes in each case. This seems to be the most straightforward way to do this assignment.\n",
    "\n",
    "When you do find a stop, you then consider the start(s) that you found along the way in this ORF.  Notice that the \"longest\" ORF is then defined by the start that is at the beginning of your start list.  Don't forget to clean up this list after you find genes using this stop since it would terminate all of those \"pending\" genes. What would happen if you didn't clean up your start list - you would be finding ORF candidates that had a stop in the middle of them, which means they are not __OPEN__ reading frames.\n",
    "\n",
    "## Termini and gene fragments\n",
    "\n",
    "This is a bit easier if you think that position 0(1 externally) will allways be a start position for a gene in every frame. This is the case where a start codon exists somewhere upstream of our sequence. It would code for a gene that would terminate in our sequence somewhere. This is the classic example of a dangling-stop. We would report this gene fragment beginning at position 0(1) and ending at the end of that in-frame stop. In frame 1, there might actually be a start codon at this position so dont count it twice. In Frames 2 or 3, there are extra bases ahead of our first full codon, so we will always have the possibility that the actual start is upstream of this dna fragment, so.. we report this as a dangling stop case starting at position 0(1).  \n",
    "\n",
    "Dangling starts always end at the end of sequence, even if that position is out of frame. In cases where no start or stop codon is found anywhere in the frame, the entire sequence is then considered as a gene.\n",
    "\n",
    "# Data and output files\n",
    "\n",
    "The cell below is set up for the non-extra credit version of the program with a fixed filename. Your much more usable program will use STDIN and STDOUT in command line fashion. To do this, simplify main so your fastaReader object is reading from STDIN.\n",
    "\n",
    "All of the input files and programs should be in the same folder so you wonâ€™t have to specify a path, if your data file is in some other folder, you can provide that like this:<br> ~/Desktop/bme160/someOtherPlace/labData.fa.\n",
    "\n",
    "# Comments\n",
    "\n",
    "To get full credit for this lab, you must turn in your design for each program you submit. Remember that any item you write or include must have proper docstrings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sequenceAnalysis import FastAreader as FastaReader\n",
    "#Pseudo Code for OrfFinder:\n",
    "#for loop that iterates through 3 frames (0 to 2):\n",
    "#for loop that iterates through codons\n",
    "#if start codon is encountered, add it to empty lists of start codons and dangling starts\n",
    "#if a stop codon is encountered:\n",
    "    #if there are no start codons in startCodonPositions and there are no ORFs in self.allORFs, it is a dangling stop. Add the orf to self.allORFs\n",
    "    #if there are no start codons in startCodonPositions, but there are ORFs in self.allORFs, then it is a arbitrary stop codon\n",
    "#if there is a start codon found before the stop codon, this segment of the sequence can be stored as an ORF. A list of the start codon pos., stop pos., and length is added to self.allORFs\n",
    "#danglingStarts and startCodonPositions are set to empty lists after the ORF is added to self.allORFs\n",
    "#if danglingStarts contains a start codon outside of the inner for loop, a dangling start has been found. The dangling start ORF is added to self.allORFs\n",
    "#Now, we iterated through the reverse strand. We use the same nested for loop, but change the way in which the start and stop codon positions are added to self.allORFs.\n",
    "#Since we are iterating through the reverse complement, start and stop codon positions must be stored according to the forward strand\n",
    "#start position of ORF = (len(reverse sequence) - stop codon pos) + 1.\n",
    "#end position of ORF = (len(reverse sequence) - start codon pos) + 1.\n",
    "\n",
    "class OrfFinder:\n",
    "\n",
    "    '''This class searches the seqeunce for ORFs. It identifies ORFs longer than or equal to 100 nucleotides.\n",
    "    It identifies dangling stop and start ORFs. It contains a method which generates the reverse complement\n",
    "    and searches for ORFs within the reverse complement, while storing positions in accordance to the first\n",
    "    strand. ORF fram, start pos., end pos., length are stored in a nested list which is returned by the findORFs\n",
    "    method.'''\n",
    "\n",
    "    def __init__(self, seq):\n",
    "        self.definedStartCodon = \"ATG\"\n",
    "        self.definedStopCodons = [\"TAG\", \"TAA\", \"TGA\"]\n",
    "        self.sequence = seq.replace(\" \",\"\")\n",
    "        self.allORFs = []\n",
    "        \n",
    "\n",
    "    def seqReverse(self):\n",
    "        '''Create reverse complement of first strand sequence.'''\n",
    "        complement = {'A':'T', 'T':'A', 'C':'G', 'G':'C'}\n",
    "        reverseComplement = \"\"\n",
    "        for nucleotide in self.sequence: \n",
    "            reverseComplement += complement.get(nucleotide)\n",
    "        return reverseComplement[::-1]\n",
    "        \n",
    "\n",
    "    def findORFs(self):\n",
    "        '''Search for ORFs, store and return nested list self.allORFs'''\n",
    "        \n",
    "        danglingStarts = []\n",
    "        reverseStarts = []\n",
    "        stopCodonPosition = 0\n",
    "        for frame in range(3):   #iterates through each frame\n",
    "\n",
    "            startIndex = frame\n",
    "            startCodonPositions = []\n",
    "\n",
    "            #print(\"checkpoint 1\")\n",
    "            for nucIndex in range(startIndex, len(self.sequence), 3):   #iterates through sequence within a frame, by 3 to enabling access to each codon\n",
    "                #print(\"checkpoint 2\")\n",
    "                codon = self.sequence[nucIndex: nucIndex + 3]   \n",
    "\n",
    "                #print(codon)\n",
    "                if codon == self.definedStartCodon:\n",
    "                    #print(\"start codon found\", codon)\n",
    "                    startCodonPositions += [nucIndex]\n",
    "                    #print(\"START CODON POSITIONS:\", startCodonPositions)\n",
    "\n",
    "                elif codon in self.definedStopCodons:   \n",
    "\n",
    "                    #print(\"stop codon found\", codon)\n",
    "                    if not startCodonPositions:\n",
    "\n",
    "                        if not self.allORFs:   #identify dangling stops\n",
    "                            #print(\"no orfs\", self.allORFs)\n",
    "                            stopCodonPosition = nucIndex + 2\n",
    "                            lengthofORF = stopCodonPosition\n",
    "                            self.allORFs += [[frame + 1, 1, stopCodonPosition + 1, lengthofORF + 1]]\n",
    "                            continue\n",
    "\n",
    "                        else:   #arbitrary stop codon encountered\n",
    "                            #print(\"list is empty\")\n",
    "                            continue\n",
    "\n",
    "                    stopCodonPosition = nucIndex + 2\n",
    "                    \n",
    "                    #print(\"debug\", startCodonPositions)\n",
    "\n",
    "                    lengthofORF = stopCodonPosition - startCodonPositions[0]\n",
    "                    self.allORFs += [[frame + 1, startCodonPositions[0] + 1, stopCodonPosition + 1, lengthofORF + 1]]\n",
    "                    \n",
    "                    startCodonPositions = [] #ORF was found, so reset startCodonPositions and danglingStarts\n",
    "                    danglingStarts = []\n",
    "                    #print(\"dangling starts cleared\")\n",
    "                \n",
    "                if startCodonPositions:\n",
    "                    danglingStarts += startCodonPositions\n",
    "                    #print(\"Dangling:\", danglingStarts)\n",
    "                #If any start codons are left at this point where frame is over, then it is a dangling start\n",
    "                \n",
    "        #print(\"list of any remaining start positions\", startCodonPositions)\n",
    "\n",
    "        \n",
    "            #print(\"Dangling s\",danglingStarts)\n",
    "            if danglingStarts: #account for dangling starts\n",
    "                stopCodonPosition = len(self.sequence) - 1\n",
    "                lengthofORF = stopCodonPosition - (danglingStarts[0] + 1)\n",
    "                self.allORFs += [[danglingStarts[0]%3 + 1, startCodonPositions[0] + 1, stopCodonPosition + 1, lengthofORF + 1]]\n",
    "            \n",
    "        #REVERSE STRAND\n",
    "        revSeq = self.seqReverse()   #store reverse strand\n",
    "        revORFs = []\n",
    "        #print(\"revseq:\", revSeq)\n",
    "        danglingStarts = []\n",
    "        for frame in range(3):\n",
    "\n",
    "            startIndex = frame\n",
    "            reverseStarts = []\n",
    "\n",
    "            #print(\"checkpoint 1\")\n",
    "            for nucIndex in range(startIndex, len(revSeq), 3):\n",
    "                #print(\"checkpoint 2\")\n",
    "                codon = revSeq[nucIndex: nucIndex + 3]\n",
    "\n",
    "                #print(codon)\n",
    "                if codon == self.definedStartCodon:\n",
    "                    #print(\"start codon found\", codon)\n",
    "                    reverseStarts += [nucIndex]\n",
    "                    #print(\"START CODON POSITIONS:\", reverseStarts)\n",
    "\n",
    "                elif codon in self.definedStopCodons:\n",
    "\n",
    "                    #print(\"stop codon found\", codon)\n",
    "                    if not reverseStarts:\n",
    "                        #print(\"no start codons found\")\n",
    "\n",
    "                        if not revORFs:   #no ORFs found yet, account for dangling stop\n",
    "                            #print(\"no orfs\", revORFs)\n",
    "                            stopCodonPosition = nucIndex + 2\n",
    "                            lengthofORF = stopCodonPosition\n",
    "                            revORFs += [[-(frame + 1),  (len(revSeq)-stopCodonPosition), len(revSeq), lengthofORF + 1]]\n",
    "                            continue\n",
    "\n",
    "                        else:   #arbitrary stop codon found\n",
    "                            #print(\"list is empty\")\n",
    "                            continue\n",
    "\n",
    "                    stopCodonPosition = nucIndex + 2\n",
    "                    \n",
    "                    #print(\"debug\", reverseStarts)\n",
    "\n",
    "                    lengthofORF = stopCodonPosition - reverseStarts[0]\n",
    "                    \n",
    "                    start = len(revSeq) - stopCodonPosition\n",
    "                    end = len(revSeq) - reverseStarts[0]\n",
    "                    #print(\"START\", start)\n",
    "                    #print(\"STOP\", end)\n",
    "\n",
    "                    revORFs += [[-(frame + 1), start, end, lengthofORF + 1]]\n",
    "                    \n",
    "                    reverseStarts = []\n",
    "                    danglingStarts = []\n",
    "                \n",
    "                if reverseStarts:\n",
    "                    danglingStarts += reverseStarts\n",
    "                #If any start codons are left at this point where frame is over, then it is a dangling start\n",
    "                \n",
    "        #print(\"list of any remaining start positions\", reverseStarts)\n",
    "\n",
    "        \n",
    "            #print(\"Dangling s\",danglingStarts)\n",
    "            if danglingStarts:   #account for dangling starts\n",
    "                stopCodonPosition = len(self.sequence)\n",
    "                lengthofORF = stopCodonPosition - (danglingStarts[0] + 1)\n",
    "                revORFs += [[-(danglingStarts[0]%3 + 1), 1, (len(revSeq) - danglingStarts[0]), lengthofORF + 1]]\n",
    "            \n",
    "        self.allORFs += revORFs\n",
    "\n",
    "        return self.allORFs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 57166..61908   4743\n",
      "-1 8192..11422   3231\n",
      "2 65963..69004   3042\n",
      "-3 14589..16862   2274\n",
      "-2 2968..4872   1905\n",
      "1 64093..65952   1860\n",
      "-3 30..1694   1665\n",
      "1 69475..71052   1578\n",
      "1 48805..50223   1419\n",
      "1 47398..48798   1401\n",
      "-3 29133..30500   1368\n",
      "-1 40922..42250   1329\n",
      "-2 19975..21270   1296\n",
      "3 72273..73559   1287\n",
      "-1 24482..25639   1158\n",
      "-1 6689..7804   1116\n",
      "1 50251..51366   1116\n",
      "-3 27501..28601   1101\n",
      "2 63038..64078   1041\n",
      "3 62019..63038   1020\n",
      "-2 42271..43263   993\n",
      "3 51864..52805   942\n",
      "1 45484..46371   888\n",
      "-3 11433..12317   885\n",
      "2 74357..75184   828\n",
      "2 71576..72394   819\n",
      "3 46341..47138   798\n",
      "-2 18613..19407   795\n",
      "1 55642..56388   747\n",
      "-1 16940..17632   693\n",
      "-2 13288..13974   687\n",
      "-3 26115..26801   687\n",
      "-1 21338..21994   657\n",
      "-1 30998..31654   657\n",
      "-2 12601..13251   651\n",
      "-2 4894..5532   639\n",
      "-3 32592..33221   630\n",
      "-1 39914..40525   612\n",
      "1 53977..54588   612\n",
      "3 54588..55193   606\n",
      "-3 33234..33809   576\n",
      "-3 22002..22559   558\n",
      "-3 23859..24413   555\n",
      "-2 1945..2490   546\n",
      "1 73861..74370   510\n",
      "-2 6214..6696   483\n",
      "1 16324..16806   483\n",
      "-1 22556..23032   477\n",
      "3 32808..33260   453\n",
      "2 53324..53776   453\n",
      "1 29056..29502   447\n",
      "-2 36286..36729   444\n",
      "3 51396..51833   438\n",
      "2 55196..55627   432\n",
      "-1 2468..2896   429\n",
      "-2 31798..32220   423\n",
      "1 52891..53307   417\n",
      "-2 30595..30996   402\n",
      "-2 5809..6204   396\n",
      "-2 35740..36129   390\n",
      "-3 34542..34925   384\n",
      "2 56474..56857   384\n",
      "-1 17888..18268   381\n",
      "-3 23004..23381   378\n",
      "-3 52968..53336   369\n",
      "-1 32207..32572   366\n",
      "-1 23396..23752   357\n",
      "2 36419..36775   357\n",
      "-2 18268..18609   342\n",
      "2 44594..44935   342\n",
      "1 43714..44049   336\n",
      "2 6557..6886   330\n",
      "-2 34927..35253   327\n",
      "-3 34131..34451   321\n",
      "1 30778..31095   318\n",
      "2 24974..25288   315\n",
      "1 1246..1557   312\n",
      "-3 35253..35561   309\n",
      "-2 39160..39468   309\n",
      "-2 68932..69234   303\n",
      "1 42430..42726   297\n",
      "2 3227..3517   291\n",
      "2 11174..11464   291\n",
      "-1 19670..19960   291\n",
      "-1 25697..25981   285\n",
      "2 13802..14080   279\n",
      "2 71114..71392   279\n",
      "2 73559..73837   279\n",
      "1 15268..15543   276\n",
      "-3 33825..34097   273\n",
      "1 28399..28668   270\n",
      "-2 39712..39981   270\n",
      "-3 27024..27290   267\n",
      "3 44052..44318   267\n",
      "-1 5546..5809   264\n",
      "-3 62670..62933   264\n",
      "-2 64780..65043   264\n",
      "3 8712..8966   255\n",
      "1 29845..30099   255\n",
      "2 30086..30337   252\n",
      "-1 39452..39703   252\n",
      "2 13373..13621   249\n",
      "-3 57690..57938   249\n",
      "3 23328..23567   240\n",
      "-1 54542..54778   237\n",
      "-3 55617..55853   237\n",
      "-1 63689..63925   237\n",
      "2 4376..4606   231\n",
      "2 60134..60364   231\n",
      "3 67791..68021   231\n",
      "3 53757..53984   228\n",
      "-3 58956..59183   228\n",
      "1 169..393   225\n",
      "3 40401..40622   222\n",
      "-1 28661..28876   216\n",
      "-1 35267..35482   216\n",
      "-3 60423..60635   213\n",
      "3 42111..42320   210\n",
      "2 18800..19006   207\n",
      "-1 53732..53938   207\n",
      "2 21965..22168   204\n",
      "-2 21694..21894   201\n",
      "3 37593..37790   198\n",
      "-1 60662..60859   198\n",
      "-3 65520..65714   195\n",
      "-2 75055..75249   195\n",
      "-1 51704..51889   186\n",
      "-1 15194..15376   183\n",
      "-1 30026..30208   183\n",
      "-3 40539..40721   183\n",
      "-3 45456..45638   183\n",
      "3 22515..22694   180\n",
      "3 33492..33671   180\n",
      "-3 47043..47222   180\n",
      "1 14098..14274   177\n",
      "-3 68733..68909   177\n",
      "2 10493..10666   174\n",
      "1 13618..13791   174\n",
      "-3 13959..14132   174\n",
      "-3 65904..66077   174\n",
      "3 68604..68777   174\n",
      "1 19519..19689   171\n",
      "-2 66178..66348   171\n",
      "-2 67264..67434   171\n",
      "1 69004..69174   171\n",
      "3 2922..3089   168\n",
      "-3 4137..4304   168\n",
      "1 23239..23406   168\n",
      "-2 35539..35706   168\n",
      "-3 40761..40928   168\n",
      "1 3157..3321   165\n",
      "1 14953..15117   165\n",
      "2 43103..43267   165\n",
      "-1 61970..62134   165\n",
      "-1 70640..70804   165\n",
      "3 3693..3854   162\n",
      "2 7400..7561   162\n",
      "2 18479..18640   162\n",
      "-3 38967..39128   162\n",
      "-3 62040..62201   162\n",
      "2 64967..65128   162\n",
      "-1 27722..27880   159\n",
      "2 37847..38005   159\n",
      "2 42767..42925   159\n",
      "1 52603..52761   159\n",
      "2 69305..69460   156\n",
      "2 6386..6538   153\n",
      "1 17569..17721   153\n",
      "1 38284..38436   153\n",
      "-3 50115..50267   153\n",
      "-3 72660..72812   153\n",
      "-3 2295..2444   150\n",
      "1 6223..6372   150\n",
      "1 26455..26604   150\n",
      "3 32571..32720   150\n",
      "-1 37655..37804   150\n",
      "-3 49050..49199   150\n",
      "-2 60016..60165   150\n",
      "-2 62347..62496   150\n",
      "-1 11978..12124   147\n",
      "3 16077..16223   147\n",
      "2 26102..26248   147\n",
      "-2 38452..38598   147\n",
      "-2 43612..43758   147\n",
      "-3 62301..62447   147\n",
      "-2 22600..22743   144\n",
      "2 28892..29035   144\n",
      "3 34758..34901   144\n",
      "-1 54191..54334   144\n",
      "2 70340..70483   144\n",
      "-3 72360..72503   144\n",
      "2 2657..2797   141\n",
      "3 17373..17513   141\n",
      "2 20747..20887   141\n",
      "2 41609..41749   141\n",
      "-1 46871..47011   141\n",
      "-3 69981..70121   141\n",
      "1 71392..71532   141\n",
      "2 4628..4765   138\n",
      "-1 7892..8029   138\n",
      "2 8981..9118   138\n",
      "2 11813..11950   138\n",
      "-3 53925..54062   138\n",
      "3 960..1094   135\n",
      "-3 19419..19553   135\n",
      "-2 25447..25581   135\n",
      "1 28066..28200   135\n",
      "3 35337..35471   135\n",
      "1 40831..40965   135\n",
      "-3 45039..45173   135\n",
      "-1 46574..46708   135\n",
      "3 72009..72143   135\n",
      "1 22303..22434   132\n",
      "-3 37449..37580   132\n",
      "-1 62957..63088   132\n",
      "-3 71487..71618   132\n",
      "-2 9241..9369   129\n",
      "-2 32416..32544   129\n",
      "-1 34139..34267   129\n",
      "1 42268..42396   129\n",
      "-2 56662..56790   129\n",
      "2 61220..61348   129\n",
      "-2 61810..61938   129\n",
      "2 65141..65269   129\n",
      "1 4153..4278   126\n",
      "2 17741..17866   126\n",
      "2 35897..36022   126\n",
      "3 48690..48815   126\n",
      "-2 66505..66630   126\n",
      "1 12226..12348   123\n",
      "-1 32957..33079   123\n",
      "-1 61496..61618   123\n",
      "2 65810..65932   123\n",
      "3 74601..74723   123\n",
      "-3 8100..8219   120\n",
      "1 14473..14592   120\n",
      "-1 15401..15520   120\n",
      "1 16936..17055   120\n",
      "3 41415..41534   120\n",
      "-3 48270..48389   120\n",
      "-2 55177..55296   120\n",
      "2 59582..59701   120\n",
      "-2 64537..64656   120\n",
      "3 69237..69356   120\n",
      "2 3851..3967   117\n",
      "1 5866..5982   117\n",
      "-2 7570..7686   117\n",
      "3 11898..12014   117\n",
      "-2 19696..19812   117\n",
      "3 21363..21479   117\n",
      "-1 29633..29749   117\n",
      "2 37442..37558   117\n",
      "-2 44371..44487   117\n",
      "1 2506..2619   114\n",
      "2 12320..12433   114\n",
      "-2 12436..12549   114\n",
      "2 31127..31240   114\n",
      "3 67479..67592   114\n",
      "2 19733..19843   111\n",
      "3 26805..26915   111\n",
      "-2 38848..38958   111\n",
      "2 41384..41494   111\n",
      "2 57992..58102   111\n",
      "2 64628..64738   111\n",
      "2 70226..70336   111\n",
      "-2 74182..74292   111\n",
      "3 4455..4562   108\n",
      "2 14561..14668   108\n",
      "-1 14576..14683   108\n",
      "1 16123..16230   108\n",
      "-2 28786..28893   108\n",
      "3 31188..31295   108\n",
      "3 37356..37463   108\n",
      "-2 41479..41586   108\n",
      "-2 47809..47916   108\n",
      "2 53984..54091   108\n",
      "-3 63294..63401   108\n",
      "-2 63634..63741   108\n",
      "-3 66393..66500   108\n",
      "-3 71112..71219   108\n",
      "-2 71560..71667   108\n",
      "-1 74795..74902   108\n",
      "1 6583..6687   105\n",
      "3 8376..8480   105\n",
      "2 9140..9244   105\n",
      "3 12366..12470   105\n",
      "-1 14444..14548   105\n",
      "-1 15632..15736   105\n",
      "3 23931..24035   105\n",
      "2 40919..41023   105\n",
      "-2 43900..44004   105\n",
      "2 59228..59332   105\n",
      "2 2414..2515   102\n",
      "1 7843..7944   102\n",
      "3 9552..9653   102\n",
      "-1 12185..12286   102\n",
      "2 13205..13306   102\n",
      "1 25858..25959   102\n",
      "-2 26872..26973   102\n",
      "3 26922..27023   102\n",
      "3 34290..34391   102\n",
      "3 35061..35162   102\n",
      "2 38267..38368   102\n",
      "1 38899..39000   102\n",
      "3 43134..43235   102\n",
      "-2 53434..53535   102\n",
      "2 64526..64627   102\n",
      "-2 65737..65838   102\n",
      "-1 69392..69493   102\n",
      "-1 72809..72910   102\n"
     ]
    }
   ],
   "source": [
    "########################################################################\n",
    "# CommandLine\n",
    "########################################################################\n",
    "class CommandLine() :\n",
    "    '''\n",
    "    Handle the command line, usage and help requests.\n",
    "\n",
    "    CommandLine uses argparse, now standard in 2.7 and beyond. \n",
    "    it implements a standard command line argument parser with various argument options,\n",
    "    a standard usage and help.\n",
    "\n",
    "    attributes:\n",
    "    all arguments received from the commandline using .add_argument will be\n",
    "    avalable within the .args attribute of object instantiated from CommandLine.\n",
    "    For example, if myCommandLine is an object of the class, and requiredbool was\n",
    "    set as an option using add_argument, then myCommandLine.args.requiredbool will\n",
    "    name that option.\n",
    " \n",
    "    '''\n",
    "    \n",
    "    def __init__(self, inOpts=None) :\n",
    "        '''\n",
    "        Implement a parser to interpret the command line argv string using argparse.\n",
    "        '''\n",
    "        \n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(description = 'Program prolog - a brief description of what this thing does', \n",
    "                                             epilog = 'Program epilog - some other stuff you feel compelled to say', \n",
    "                                             add_help = True, #default is True \n",
    "                                             prefix_chars = '-', \n",
    "                                             usage = '%(prog)s [options] -option1[default] <input >output'\n",
    "                                             )\n",
    "        self.parser.add_argument('-lG', '--longestGene', action = 'store', nargs='?', const=True, default=False, help='longest Gene in an ORF')\n",
    "        self.parser.add_argument('-mG', '--minGene', type=int, choices= (100,200,300,500,1000), default=100, action = 'store', help='minimum Gene length')\n",
    "        self.parser.add_argument('-s', '--start', action = 'append', default = ['ATG'],nargs='?', \n",
    "                                 help='start Codon') #allows multiple list options\n",
    "        self.parser.add_argument('-t', '--stop', action = 'append', default = ['TAG','TGA','TAA'],nargs='?', help='stop Codon') #allows multiple list options\n",
    "        self.parser.add_argument('-v', '--version', action='version', version='%(prog)s 0.1')  \n",
    "        if inOpts is None :\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "########################################################################\n",
    "# Main\n",
    "########################################################################\n",
    "   \n",
    "def main(inFile = None, options = None):\n",
    "    '''\n",
    "    Create an instance of class OrfFinder, generate a nested list containing lists of the frame, start position, \n",
    "    stop position, and frame. Sort the ORFs in descending order of length, and print in the output file given by\n",
    "    CommandLine. \n",
    "    '''\n",
    "    thisCommandLine = CommandLine(options) #instantiate commandLine object with parameters(options)\n",
    "    reader = FastaReader(inFile)\n",
    "    \n",
    "    for header, sequence in reader.readFasta():\n",
    "        nucleotideSequence = OrfFinder(sequence)\n",
    "        geneList = sorted(nucleotideSequence.findORFs(), key=lambda x:(-x[3], x[1])) #Sorted by descending length\n",
    "        for orfList in geneList: \n",
    "            #print(orfList)\n",
    "            if orfList[3] >= 100:\n",
    "                print(f'{orfList[0]} {orfList[1]}..{orfList[2]}   {orfList[3]}')\n",
    "    \n",
    "if __name__ == \"__main__\":\n",
    "    #delete this stuff if running from commandline\n",
    "    main(inFile = 'tass2.fa', options = ['-mG=300', '-lG'])    \n",
    "    #main()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Inspections"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hema Prasanna: The code looks good to me. I recommend that you remove all the commented out print statements to make the code more concise and easier to read. I think it would be useful to add more comments to explain how you are calculting start and stop codon positions for the reverse strand. Other than that, I think that you did a great job1\n",
    "\n",
    "Akash Pandit: Everything runs fine! In the future you could utilize more helper methods instead of having one method to carry out ORF finding for the forward and the reverse strand"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python3 findOrfs.py  <somefile  >someOtherFile -mG 300 -lG -s ATG -s GTG"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findORFs output using tass2.fa  -mG=300 -lG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tass2 NODE_159_length_75728_cov_97.549133\n",
    "+1 57166..61908  4743\n",
    "-1  8192..11422  3231\n",
    "+2 65963..69004  3042\n",
    "-3 14589..16862  2274\n",
    "-2  2968.. 4872  1905\n",
    "+1 64093..65952  1860\n",
    "-3    30.. 1694  1665\n",
    "+1 69475..71052  1578\n",
    "+1 48805..50223  1419\n",
    "+1 47398..48798  1401\n",
    "-3 29133..30500  1368\n",
    "-1 40922..42250  1329\n",
    "-2 19975..21270  1296\n",
    "+3 72273..73559  1287\n",
    "-1 24482..25639  1158\n",
    "-1  6689.. 7804  1116\n",
    "+1 50251..51366  1116\n",
    "-3 27501..28601  1101\n",
    "+2 63038..64078  1041\n",
    "+3 62019..63038  1020\n",
    "-2 42271..43263   993\n",
    "+3 51864..52805   942\n",
    "+1 45484..46371   888\n",
    "-3 11433..12317   885\n",
    "+2 74357..75184   828\n",
    "+2 71576..72394   819\n",
    "+3 46341..47138   798\n",
    "-2 18613..19407   795\n",
    "+1 55642..56388   747\n",
    "-1 16940..17632   693\n",
    "-2 13288..13974   687\n",
    "-3 26115..26801   687\n",
    "-1 21338..21994   657\n",
    "-1 30998..31654   657\n",
    "-2 12601..13251   651\n",
    "-2  4894.. 5532   639\n",
    "-3 32592..33221   630\n",
    "-1 39914..40525   612\n",
    "+1 53977..54588   612\n",
    "+3 54588..55193   606\n",
    "-3 33234..33809   576\n",
    "-3 22002..22559   558\n",
    "-3 23859..24413   555\n",
    "-2  1945.. 2490   546\n",
    "+1 73861..74370   510\n",
    "-2  6214.. 6696   483\n",
    "+1 16324..16806   483\n",
    "-1 22556..23032   477\n",
    "+3 32808..33260   453\n",
    "+2 53324..53776   453\n",
    "+1 29056..29502   447\n",
    "-2 36286..36729   444\n",
    "+3 51396..51833   438\n",
    "+2 55196..55627   432\n",
    "-1  2468.. 2896   429\n",
    "-2 31798..32220   423\n",
    "+1 52891..53307   417\n",
    "-2 30595..30996   402\n",
    "-2  5809.. 6204   396\n",
    "+1     1..  393   393\n",
    "-2 35740..36129   390\n",
    "-3 34542..34925   384\n",
    "+2 56474..56857   384\n",
    "-1 17888..18268   381\n",
    "-3 23004..23381   378\n",
    "-3 52968..53336   369\n",
    "-1 32207..32572   366\n",
    "-1 23396..23752   357\n",
    "+2 36419..36775   357\n",
    "-2 18268..18609   342\n",
    "+2 44594..44935   342\n",
    "+1 43714..44049   336\n",
    "+2  6557.. 6886   330\n",
    "-2 34927..35253   327\n",
    "-3 34131..34451   321\n",
    "+1 30778..31095   318\n",
    "+2 24974..25288   315\n",
    "+1  1246.. 1557   312\n",
    "-3 35253..35561   309\n",
    "-2 39160..39468   309\n",
    "-2 68932..69234   303"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# findORFs output using lab5test.fa at mG=0 cutoff\n",
    "\n",
    "(Note: I have included the sequence of the gene candidate as an additional column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test\n",
    "+1     1..    9     9 ATGATGTAA\n",
    "+3     1..    9     9 ATGATGTAA\n",
    "-1     1..    9     9 TTACATCAT\n",
    "-2     1..    9     9 TTACATCAT\n",
    "-3     1..    9     9 TTACATCAT\n",
    "+1     4..    9     6 ATGTAA\n",
    "+2     1..    4     4 ATGA\n",
    "test2\n",
    "+1     1..   10    10 AATGATGTAA\n",
    "+2     1..   10    10 AATGATGTAA\n",
    "-1     1..   10    10 TTACATCATT\n",
    "-2     1..   10    10 TTACATCATT\n",
    "-3     1..   10    10 TTACATCATT\n",
    "+2     2..   10     9 ATGATGTAA\n",
    "+2     5..   10     6 ATGTAA\n",
    "+3     1..    5     5 AATGA\n",
    "test3\n",
    "+2     1..   11    11 AAATGATGTAA\n",
    "+3     1..   11    11 AAATGATGTAA\n",
    "-1     1..   11    11 TTACATCATTT\n",
    "-2     1..   11    11 TTACATCATTT\n",
    "-3     1..   11    11 TTACATCATTT\n",
    "+3     3..   11     9 ATGATGTAA\n",
    "+1     1..    6     6 AAATGA\n",
    "+3     6..   11     6 ATGTAA\n",
    "test-1\n",
    "+1     1..    9     9 TTACATCAT\n",
    "+2     1..    9     9 TTACATCAT\n",
    "+3     1..    9     9 TTACATCAT\n",
    "-1     1..    9     9 ATGATGTAA\n",
    "-3     1..    9     9 ATGATGTAA\n",
    "-1     1..    6     6 ATGTAA\n",
    "-2     6..    9     4 ATGA\n",
    "test-2\n",
    "+1     1..   10    10 TTACATCATA\n",
    "+2     1..   10    10 TTACATCATA\n",
    "+3     1..   10    10 TTACATCATA\n",
    "-1     1..   10    10 TATGATGTAA\n",
    "-2     1..   10    10 TATGATGTAA\n",
    "-2     1..    9     9 ATGATGTAA\n",
    "-2     1..    6     6 ATGTAA\n",
    "-3     6..   10     5 TATGA\n",
    "test-3\n",
    "+1     1..   11    11 TTACATCATAA\n",
    "+2     1..   11    11 TTACATCATAA\n",
    "+3     1..   11    11 TTACATCATAA\n",
    "-2     1..   11    11 TTATGATGTAA\n",
    "-3     1..   11    11 TTATGATGTAA\n",
    "-3     1..    9     9 ATGATGTAA\n",
    "-3     1..    6     6 ATGTAA\n",
    "-1     6..   11     6 TTATGA\n",
    "test1A\n",
    "+3     1..   10    10 ATGATGTAAA\n",
    "-1     1..   10    10 TTTACATCAT\n",
    "-2     1..   10    10 TTTACATCAT\n",
    "-3     1..   10    10 TTTACATCAT\n",
    "+1     1..    9     9 ATGATGTAA\n",
    "+1     4..    9     6 ATGTAA\n",
    "+2     1..    4     4 ATGA\n",
    "test2A\n",
    "+1     1..   11    11 AATGATGTAAA\n",
    "-1     1..   11    11 TTTACATCATT\n",
    "-2     1..   11    11 TTTACATCATT\n",
    "-3     1..   11    11 TTTACATCATT\n",
    "+2     1..   10    10 AATGATGTAA\n",
    "+2     2..   10     9 ATGATGTAA\n",
    "+2     5..   10     6 ATGTAA\n",
    "+3     1..    5     5 AATGA\n",
    "test3A\n",
    "+2     1..   12    12 AAATGATGTAAA\n",
    "-1     1..   12    12 TTTACATCATTT\n",
    "-2     1..   12    12 TTTACATCATTT\n",
    "-3     1..   12    12 TTTACATCATTT\n",
    "+3     1..   11    11 AAATGATGTAA\n",
    "+3     3..   11     9 ATGATGTAA\n",
    "+1     1..    6     6 AAATGA\n",
    "+3     6..   11     6 ATGTAA\n",
    "test-1A\n",
    "+1     1..   10    10 ATTACATCAT\n",
    "+2     1..   10    10 ATTACATCAT\n",
    "+3     1..   10    10 ATTACATCAT\n",
    "-3     1..   10    10 ATGATGTAAT\n",
    "-1     2..   10     9 ATGATGTAA\n",
    "-1     2..    7     6 ATGTAA\n",
    "-2     7..   10     4 ATGA\n",
    "test-2A\n",
    "+1     1..   12    12 AATTACATCATA\n",
    "+2     1..   12    12 AATTACATCATA\n",
    "+3     1..   12    12 AATTACATCATA\n",
    "-1     1..   12    12 TATGATGTAATT\n",
    "-2     3..   12    10 TATGATGTAA\n",
    "-2     3..   11     9 ATGATGTAA\n",
    "-2     3..    8     6 ATGTAA\n",
    "-3     8..   12     5 TATGA\n",
    "test-3A\n",
    "+1     1..   13    13 AATTACATCATAA\n",
    "+2     1..   13    13 AATTACATCATAA\n",
    "+3     1..   13    13 AATTACATCATAA\n",
    "-2     1..   13    13 TTATGATGTAATT\n",
    "-3     3..   13    11 TTATGATGTAA\n",
    "-3     3..   11     9 ATGATGTAA\n",
    "-3     3..    8     6 ATGTAA\n",
    "-1     8..   13     6 TTATGA"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
